Chore/reporting (#643)
OWASP/wstg: document/5-Reporting/README.md

## About this Section

This guide provides only suggestions about one possible approach to reporting, and should not be treated as as strict rules that must be followed. When considering any of the recommendations below, always ask yourself whether the recommendation would improve your report.

This guide to reporting is a best fit for consultancy-based reports. It may be overkill for internal or bug bounty reports.

Regardless of the audience, it's advisable to secure the report and encrypt it to ensure that only the receiving party is able to use it.

A good report helps your client understand your findings and highlights the quality of your technical testing. The quality of the technical testing is completely irrelevant if the client can't understand your findings.

## 1. Introduction

### 1.1 Version Control

Sets report changes, mostly presented in a table format such as the below.

| Version | Description | Date | Author |
|:-------:|-------------|------|--------|

> | 1.0 | Initial report | DD/MM/YYYY | Foo Bar |

Reviewer: ```suggestion
| 1.0 | Initial report | DD/MM/YYYY | J. Doe |
```

## 3. Findings

This section is aimed at the technical team. It should include all the necessary information to understand the vulnerability, replicate it, and resolve it. Logical separation can help improve the readability of the report. For example, you might have separate sections titled "External Access" and "Internal Access".

If this is a re-test, you might create a subsection that summarizes findings of the previous test, the updated status of previously identified vulnerabilities, and any cross-references with the current test.

### 3.1 Findings Summary

A list of the findings with their risk level. A table can be used for ease of use by both teams.

| Ref. ID |  Title | Risk Level |
|:------------:|--------|------------|
| 1 | User Authentication Bypass | High |

### 3.2 Findings Details

Each finding should be detailed with the following information:

- Reference ID, which can be used for communication between parties and for cross-references across the report.
- The vulnerability title, such as "User Authentication Bypass".
- The likelihood or exploitability of the issue, based on various factors such as:

>   - How easy it is to exploit it.

Reviewer: ```suggestion
  - How easy it is to exploit.
```

>   - Whether there is a working exploit code for it.

Reviewer: ```suggestion
  - Whether there is working exploit code for it.
```

> Performing the technical side of the assessment is only half of the overall assessment process. The final product is the production of a well written and informative report. A report should be easy to understand and should highlight all the risks found during the assessment phase. The report should appeal to both executive management and technical staff.

Reviewer 2: Reporting is very often overlooked, so might add a bold sentence like:

> **The quality of the technical testing is completely irrelevant if the client can't understand your findings.**

But up to you.

> Note: It is advised to secure the report and encrypt it to ensure that only the receiving party is able to use it.

Reviewer 2: I might also add a sentence after this along the lines of:

> This guide provides only suggestions about one possible approach to reporting, and should not be treated as as strict rules that most be followed. When considering any of the recommendations below, always ask yourself: *does this make my report better?*

## 1. Introduction

### 1.1 Version Control

Sets report changes, mostly presented in a table format such as the below.

| Version | Description | Date | Author |
|:-------:|-------------|------|--------|
| 1.0 | Initial report | DD/MM/YYYY | Foo Bar |

### 1.2 Table of Contents

A table of contents page for the document.

### 1.3 The Team

A list of the team members detailing their expertise and qualifications.

### 1.4 Scope

The boundaries and the needs of the engagement agreed upon with the organization.

### 1.5 Limitations

Limitations can be:

- Areas that were off boundaries that related to the test.
- Broken functionality.
- Lack of cooperation.
- Lack of time.

> - Lack of accesses.

Reviewer 2: "Lack of access" or "Lack of access or credentials"

### 1.6 Timeline

The duration of the engagement.

### 1.7 Disclaimer

> A disclaimer providing legal protection. The below example should be modified to your use case and shouldn't be used as is.

Reviewer 2: Bold this warning - we don't want anyone claiming this is legal advice we're giving

> *This test is a "point in time" assessment and as such the environment could have changed since the test was run. There is no guarantee that all possible security issues have been identified, and that new vulnerabilities may have been discovered since the tests were run. Thus, this report serves as a guiding document and not a warranty that the report provides a full representation of the risks threatening the systems at hand.*

Reviewer 2: Maybe replace "thus" with "As such" ?

Performing the technical side of the assessment is only half of the overall assessment process. The final product is the production of a well written and informative report. A report should be easy to understand and should highlight all the risks found during the assessment phase. The report should appeal to both executive management and technical staff.

> **The quality of the technical testing is completely irrelevant if the client can't understand your findings.**

Reviewer 3: ```suggestion
A good report helps your client understand your findings and highlights the quality of your technical testing.
```

*This test is a "point in time" assessment and as such the environment could have changed since the test was run. There is no guarantee that all possible security issues have been identified, and that new vulnerabilities may have been discovered since the tests were run. As such, this report serves as a guiding document and not a warranty that the report provides a full representation of the risks threatening the systems at hand.*

## 2. Executive Summary

This is like the elevator pitch of the report, it aims at providing executives with:

- The objective of the test, which contains the need for it and the answers that the organization need to better understand their systems.
- The key findings in a business context, such as compliance issues, reputational damage, etc.
  - It shouldn't state the technical findings and the technical impact. Only discuss the business impact.
- The strategic recommendations on how the business can stop the issues from happening again.
  - It shouldn't state technical and specific recommendations, as this concerns the technical team.

> **Note:** The summary should be constructive and meaningful, not full of jargon and destructive takeaways. In case graphs are used, make sure they help deliver a message in a clearer way than text would.

Reviewer 2: I don't really agree with the change to the strategic recommendations suggested here. Mitigating the *current* issues is generally more of a tactical recommendation (such as using HTML encoding to fix the XSS).

Strategic recommendations should address the underlying root cause of the issue (insecure development practices), and give them guidance on how to avoid having the same problems in future (implementing a secure development/defensive coding methodology, using frameworks that protect against these types of issues, etc).

## 3. Findings

This section is aimed at the technical team. It should include all the necessary information to understand the vulnerability, replicate it, and resolve it. Logical separation can help improve the readability of the report (*e.g* external access, internal access, etc.)

> *If this is a re-test, a subsection could be set with the findings summary of the previous test, the status of the identified vulnerabilities, and any cross-references with the current test.*

Reviewer 3: ```suggestion
If this is a re-test, you might create a subsection that summarizes findings of the previous test, the updated status of previously identified vulnerabilities, and any cross-references with the current test.
```

### 3.1 Findings Summary

A list of the findings with their risk level. A table can be used for ease of use by both teams.

| Ref. ID |  Title | Risk Level |
|:------------:|--------|------------|
| 1 | User Authentication Bypass | High |

### 3.2 Findings Details

Each finding should be detailed with the following information:

- Reference ID, which can be used for communication between parties and for cross-references across the report.
- The vulnerability title, such as "User Authentication Bypass".
- The likelihood or exploitability of the issue, based on various factors such as:
  - How easy it is to exploit it.
  - Whether there is a working exploit code for it.
  - The level of access required.
  - Attacker motivation to exploit it.
- The impact of the vulnerability on the system.
- Risk of the vulnerability on the application.

>   - Possible values are: "Informational", "Low", "Medium", "High", "Critical".
>     - The values set in here should be detailed in an appendix to allow the reader to understand how each score was set.

Reviewer 3: ```suggestion
  - Some suggested values are: Informational, Low, Medium, High, and Critical. Ensure that you detail the values you decide to use in an appendix. This allows the reader to understand how each score is determined.
```

  - On certain engagements it is required to have a [CVSS](https://www.first.org/cvss/) score. If not required, sometimes it is good to have, and other times it just adds complexity to the report.

> - Detailed description of what the vulnerability is, how to exploit it, and the damage that may result from its abuse. Any sensitive data should be masked (*e.g.* passwords, personal information, credit card details, etc.)

Reviewer 3: ```suggestion
- Detailed description of what the vulnerability is, how to exploit it, and the damage that may result from its exploitation. Any possibly-sensitive data should be masked, for example, passwords, personal information, or credit card details.
```

- Detailed steps on how to remediate the vulnerability, possible improvements that could help strengthen the security posture, and missing security practices.
- Additional resources that could help the reader to understand the vulnerability, such as an image, a video, a CVE, an external guide, etc.

> The above can be formatted in the way the tester deems as fitting to deliver their message in the best possible way.
> 
> **Note:** Always check if the issue provides enough information for the engineer reading this report and if they can take action based on it. If not, the issue should be better explained.

Reviewer 3: ```suggestion
Format this section in a way that best delivers your message.

Always ensure that your descriptions provide enough information for the engineer reading this report to take action based on it. Explain the finding thoroughly and provide as much technical detail as might be necessary to remedy it.
```

## Appendices

Multiple appendices can be added, such as:

- Test methodology used.
- Severity and risk rating explanations.
- Relevant output from tools used.
  - Make sure to clean the output and not just dump it.
- A checklist of all the tests conducted, such as the [WSTG checklist](https://github.com/OWASP/wstg/tree/master/checklist). These can be provided as attachments to the report.

## References

> _This is not part of the report. More guidance to writing your reports._

Reviewer 3: ```suggestion
This section is not part of the suggested report format. The below links provide more guidance to writing your reports.
```

Performing the technical side of the assessment is only half of the overall assessment process. The final product is the production of a well written and informative report. A report should be easy to understand and should highlight all the risks found during the assessment phase. The report should appeal to both executive management and technical staff.

> Note: It is advised to secure the report by using a password or any similar technology as pentest reports are marked as sensitive information.

Reviewer 2: Talking about a password brings to mind the rubbish "password protection" that things like MS Access used to have, that you could trivially bypass. The key thing is that the report needs to be *encrypted*.

## 1. Executive Summary

This is like the elevator pitch of the report, it aims at providing executives with:

> - the run down of the report in a page or two.

Reviewer 2: This seems to be mostly repeating the points below - perhaps remove or turn into a few sentences?

- quick actions and their criticality.
- the status of their application.

> ### 1.1 Objective

Reviewer 2: I like to have an introduction section in the report, which contains all the boring stuff:

- Contents page
- Details of who did the testing
- Distribution list/client contacts
- Version history
- A disclaimer somewhere to state that a pentest is a point-in-time assessment, and that you're not guaranteeing to have found every issue in the application.

You could also put things like the scope/timeline in here - because an executive doesn't really care about them most of the time.

> The purpose of this engagement and the need for it.

Reviewer 2: This should include any specific questions that they were trying to answer, and also either here or the next section should include the perspective from which testing was carried out (authenticated, given a digital certificate, etc)

### 1.2 Scope

> What this test will cover precisely. This is crucial in order to set the boundaries of the engagement.

Reviewer 2: This should be in the past tense, rather than the future tense.

It's also important to list any caveats on the scope (broken functionality, areas that you weren't allowed to test, other limitations)

### 1.3 Timeline

The duration of the engagement.

### 1.4 Summary

> The summary should contain, in the best possible form to deliver the message (*e.g.* graphs, tables, lists, etc.):

Reviewer 2: If you're going to include graphs, then make sure that they're actually useful. For example, don't put a pie chart showing risk levels (which is pretty common)

> - Key risks identified by the test.

Reviewer 2: When I see bad executive summaries, it's usually because they are just trying to summarise the contents of the report. That's fine for a *technical* summary, but not an *executive* summary. For example:

> The application is vulnerable to blind SQL injection, which allows a user to read any information from the database.

Is not good for an executive summary, because a CEO doesn't know care what SQL injection, and may not even know what a database is.

The key thing about an executive summary is that it need to explain the what the findings of the report mean *in business context*. This means talking about things like compliance with PCI DSS/GDPR/HIPAA, fines, reputational damage, etc.

Not "This is what I found", but more "This is why you should care about what I found".

> - Key recommendations for the organization.

Reviewer 2: If you're giving recommendations, these should generally be strategic rather than tactical.

A tactical recommendation would be "install the missing patches" - and you can give that in the issue section.

A strategic recommendation would be more like "implement a regular patching policy, and carry out authenticated patch scans to identify any that are missing".

Not "This is how to fix the issues" (because that's explained in the issues), but "This is how to stop the issues happening again".

> - State of the application. A concise and direct conclusion that states whether the application is secure or not.

Reviewer 2: Not sure about this - can you ever state that an application is secure? Or just that you didn't find any significant issues

Author: Sometimes an application can afford to go live and have patches along the way. A misconforming error mesage doesn't stop a release, but an authentication bypass does. Let me try and reword that.

Reviewer 2: As a pentester, that's not really your decision to make. Your job is to explain the vulnerabilities that you found and how they can be exploited - it's up to the organisation to decide what those mean for them.

Author: Valid point, let me remove it.

**Note:** The summary should be constructive and meaningful, not full of jargon and destructive takeaways.

> ## 2. Detailed Findings

Reviewer 2: Somewhere in this section there should be a table of all the risks with something like:

- Reference
- Title
- Risk level

These usually get copied into Excel spreadsheets and used to track the resolution status by the devs

> This section is aimed at the technical team. It should include all the necessary information to understand the vulnerability, replicate it, and resolve it.

Reviewer 2: It can be good to split this into multiple sections - for example if you have two applications, or a webapp and a mobile app.

Each finding should be detailed with the following information:

> - Reference ID, which can be used for communication between parties and for other references across the report.

Reviewer 2: Somewhere in here you should also be cross-referencing any other relevant issues (for example, from the lack of a CSP to the application being vulnerable to XSS). I prefer to do that in the issue body, but some reports have a list of "related issues" or something like that

- The vulnerability identified in a clear form, such as "Privilege Escalation from Client to Admin".

> - Exploitability, or likelihood. This states the persona that is capable of exploiting said vulnerability based on how easy it is to achieve it, the level of access, and the impact achieved, which is related to the motivation behind an attack.

Reviewer 2: I don't think "persona" is the right word here.

Would this perhaps look better as a nested list? Something like:

* The likelihood or exploitability of the issue, based on various factors such as:
  * How easy it is to exploit.
  * Whether public exploit code is available.
  * The level of access required.
  * Whether an attacker would be motivated to exploit it

Also, "impact" isn't really part of the exploitability/likelihood - it should be its own section

>   - Possible values are "User", "Technical", "Hacker".

Reviewer 2: I disagree with this - I don't think "Exploitability: Hacker" tells you anything

- Risk of the vulnerability on the application.

>   - Possible values are: "Informational", "Low", "Medium", "High", "Critical".

Reviewer 2: What is this based on? The way I've usually see it is that you have "impact" and "likelihood", and you work out the overall risk by (effectively) multiplying those. Should be explained in an appendix

>   - On certain engagements it is required to have the [CVSS](https://www.first.org/cvss/) score. If not required, sometimes it is good to have, and other times it just adds complexity to the report.

Reviewer: ```suggestion
  - On certain engagements it is required to have a [CVSS](https://www.first.org/cvss/) score. If not required, sometimes it is good to have, and other times it just adds complexity to the report.
```

> - A detailed description of what is the vulnerability, how to exploit it, and the damage it can inflict.

Reviewer: ```suggestion
- A detailed description of what the vulnerability is, how to exploit it, and the damage that may result from its abuse.
```

> - A detailed description of what the vulnerability is, how to exploit it, and the damage that may result from its abuse.

Reviewer 2: This should include your screenshots, PoC code, etc

Should also be masking sensitive data (passwords, personal information, PANs, etc)

> - A detailed remediation on how to resolve the vulnerability, possible improvements that could help strengthen the posture, and missing security practices.

Reviewer 2: "A detailed remediation" doesn't quite read right, should be something like "Detailed steps on how to remediate the vulnerability"

> - Provide any additional resources that could help the reader to understand the vulnerability, such as an image, a video, a CVE, an external guide, etc.

Reviewer 2: Remove "Provide" - doesn't fit with the others

> Findings can be delivered in normal format, like the above, it could be delivered in a table format that can formatted in various ways, or any other format that the testers see as fitting to deliver the message in a more comprehensible format.

Reviewer 2: You use the word "format" five times in one sentence

> Findings can be delivered in normal format, like the above, it could be delivered in a table format that can formatted in various ways, or any other format that the testers see as fitting to deliver the message in a more comprehensible format.

Reviewer 2: By "normal format" I presume you mean "text"?

One question that you should always ask when writing an issue is:

> Does this give enough information for the developer to understand and fix this issue?

And if not, then you need to add something.

## 3. Conclusion

State the organization's next steps, specifying the positive and negative findings of the report, and improvements on a big picture, such as "implement unit tests across the organization to set a security baseline".

> This should be concise, providing the organization with high level points to go after, risks to consider before going live, etc.

Reviewer 2: This just sounds like it's repeating the executive summary?

> ## Appendices

Reviewer 2: Is it worth covering retesting somewhere? If you've tested the environment before, you also want a retest section that includes:

- A high-level summary of the change since the last test
- A table showing the resolution status of every issue
- Cross-references to any outstanding issues

Multiple appendices can be added, such as:

- Test methodology used.
- Severity and risk rating explanations.

> - Tools used, scripts, and the relevant generated output.

Reviewer 2: I wouldn't normally bother with this, but quite a lot of companies do. Not really that useful to find out that they used Firefox, Chrome and Notepad on a test...

  - Make sure to clean the output and not just dump it.

> - A checklist of all the tests conducted, such as the [WSTG checklist](https://github.com/OWASP/wstg/tree/master/checklist). These can provided as attachements to the report.

Author: ```suggestion
- A checklist of all the tests conducted, such as the [WSTG checklist](https://github.com/OWASP/wstg/tree/master/checklist). These can be provided as attachments to the report.
```
